File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height:List[int])->int:
        if len(height)<3:
            return 0

        volume=0
        left,right = 0,len(height)-1
        left_max,right_max=0,0

        while left<right:
            if height[left]<=height[right]:
                if height[left]<left_max:
                    volume+=left_max-height[left]
                else:
                    left_max=height[left]
                left+=1
            else:
                if height[right]<right_max:
                    volume+=right_max-height[right]
                else:
                    right_max=height[right]
                right-=1
        return volume

File: 42-trapping-rain-water/42-trapping-rain-water.dart
import 'dart:math';

class Solution {
  int trap(List<int> height) {
    if (height.isEmpty) return 0;

    int volume = 0;
    int left = 0;
    int right = height.length - 1;
    int leftMax = height[left];
    int rightMax = height[right];

    while (left < right) {
      leftMax = max(height[left], leftMax);
      rightMax = max(height[right], rightMax);

      if (leftMax <= rightMax) {
        volume += leftMax - height[left];
        left += 1;
      } else {
        volume += rightMax - height[right];
        right -= 1;
      }
    }
    return volume;
  }
}
